WIP: Parent group overview

Martchus opened the pull request:
Just to show how it looks like so far:
![spectacle z19584](https://cloud.githubusercontent.com/assets/10248953/19963960/3df07522-a1bf-11e6-9801-affc93aca877.png)

So there are no comments and tags so far. Review badges are only displayed if it is reviewed in all job groups.

This time inclusion of previous commit (present in other PR) is intended because it provides the menu to access the parent group overview.

okurz commented on lib/OpenQA/Schema/Result/JobGroupParents.pm:
```diff
@@ -101,4 +101,32 @@ around 'default_priority' => sub {
+    my @group_ids;
+    for my $group ($self->children) {
+        push(@group_ids, $group->id);
+    }
```
can you replace this with a map?

```perl
my @group_ids = map { $_->id } $self->children;
```

@coolo would call this "pseudo code", i.e. might work, not tested.

okurz commented on lib/OpenQA/Schema/Result/JobGroupParents.pm:
```diff
@@ -101,4 +101,32 @@ around 'default_priority' => sub {
+        my $m = CommentsMarkdownParser->new;
+        return Mojo::ByteStream->new($m->markdown($self->description));
+    }
+    return;
```
maybe it looks cleaner if turned around with "early return":

```perl
return unless $self->description;
my $m = CommentsMarkdownParser->new;
return Mojo::ByteStream->new($m->markdown($self->description));
```

okurz commented on t/ui/14-dashboard.t:
```diff
@@ -98,8 +98,13 @@ subtest 'filter form' => sub {
 like($driver->get_page_source(), qr({"results":\[.*{"0048":), 'page rendered as JSON');
 
+# parent group overview
+# FIXME: Add new fixtures or add parents dynamically?
```
add parents dynamically

coolo commented:
now I realize that I jumped the gun when creating parents for the opensuse.org job groups. A parent only makes sense if the job groups share a common build pattern (or are even the same product).

I guess we need to express this better. (just a random thought)

Martchus commented:
Groups for products was the reason why we wanted this feature in the first place.

But if there are no common builds, this is no problem, too. Then there's is just no benefit regarding accumulated build results.
However, it would still make sense to have a some other parent groups. Eg. a group 'Deprecated' for old groups (eg. zz-dead: Maintenance: SLE 12 GA Updates). This would make sense even if those groups don't have common builds.

coolo commented:
Yeah, I did the grouping to ease the navbar. So we can group job groups without recent builds freely, but have to be careful with job groups that could share build numbers. E.g. Tumbleweed and Tumbleweed ports is such a case where grouping them might not be what we want - even if they share the versioning scheme.

okurz commented:
@coolo remember "dark launch" of features :-) We can add a query parameter to group job groups within parent groups based on builds and release that and see all weird behaviour which we want to tackle before adding a gui element and brag about the new feature :-)

coolo commented:
No, I want the summary to be the default. The grouping can be undone if it hurts - possibly we will need an option in the job groups.

Martchus commented:
> possibly we will need an option in the job groups

You mean an option to control whether build results are combined (and combining is the default)?

> remember "dark launch" of features

Might be an option. I removed the navbar commit here (it is a different feature anyways).

So if you think the overview would be useful in the way I implemented it I'll continue writing tests.

coolo commented:
So I played with it and I wonder if the use case of this page isn't a bit narrow. It's basically only there to make screenshots - you can't do anything else here

coolo commented:
But something else: we miss a way to intersect job groups and parent groups. E.g. the Desktop product only has on job group and so a parent group doesn't make sense. But that means it's in misc - and would be below released/dead.

coolo commented:
ok, the desktop has 2 - but SAP and HA have one

coolo commented:
btw: how is the order on the / dashboard defined? So far we have Acceptance of Server and Desktop first - but if we group Server stuff as first, it should be all Server first, then Desktop. No?

Martchus commented:
> But that means it's in misc - and would be below released/dead

Which is unacceptable in my opinion. That's why I've put the 'misc' groups at the top in the first place.

> So far we have Acceptance of Server and Desktop first - but if we group Server stuff as first, it should be all Server first, then Desktop. No?

Yes, but Server and Desktop would be just one entry (without listing the nested Acceptance, ...) because builds are combined. But maybe this wouldn't be a good idea for all users - so we could put in a option to switch between current version and this proposal.

okurz commented:
and IMHO the order of precedence should be:
* custom ordering of job groups (regardless of being job groups or parent groups)
* parent groups (when not ordered in custom way)
* orphan groups (without custom order)

coolo commented:
I deployed master on openqa.suse.de and grouped the products that share builds. But the ordering is not optimal I feel

okurz commented:
It already looks pretty tidied up in comparison to before, especially including SP3

coolo commented:
I promised a mockup, here is it: https://w3.nue.suse.com/~coolo/group-proposal.png

The duplicated build number for the groups is a bit stupid, but we need 2 links - one to the group_overview and one to tests/overview. I'm thinking about putting one of them on the (the then no longer optional) badge

coolo commented:
(and yes, I screwed up the grouping a bit in the mockup)

okurz commented:
Interesting approach.

You know that we make use of '@' in "build identifiers" already, right?
I am not sure what benefit it would bring to have the build bars for e.g. "Acceptance" and "Autoyast" over each other. This way the build bars will never look "similar". You would rather compare "Build X+1" against "Build X" in total or "all builds on Acceptance" which your view makes harder to do.

To make it short: I like @Martchus approach more.

coolo commented:
@Martchus has no approach to the global dashboard yet

coolo commented:
And yes, having the bars comparable is definitely a goal. And for that we need better visual grouping of what makes up a build than I was able to with mocking.

coolo commented:
and the parent_group_overview page might be easier visually, but to avoid repeating myself: https://github.com/os-autoinst/openQA/pull/976#issuecomment-259103648

coolo commented:
I added some spacing to make the comparision easier: https://w3.nue.suse.com/~coolo/group-proposal2.png

okurz commented:
as disussed for index page:
1. folding of children group results by default
2. "child group name" below parent group links to build result of child group (not group overview page)

coolo commented:
I accepted this into a new branch - and tried to code towards the mockup. This still needs a lot of work though
